### PR TITLE
Handle Paddle billing customer already exists error

### PR DESCRIPTION
**Pull request by nduplessis**
## Pull Request

**Summary:**
When attempting to add a customer to Paddle Billing that already exists, an unhandled error occurs

**Description:**
https://developer.paddle.com/errors/customers/customer_already_exists

**Testing:**

**Screenshots (if applicable):**

**Checklist:**

- [ ] Code follows the project's coding standards
- [ ] Tests have been added or updated to cover the changes
- [ ] Documentation has been updated (if applicable)
- [ ] All existing tests pass
- [ ] Conforms to the contributing guidelines

**Additional Notes:**

### CODE REVIEW

**Comment by excid3** [comment]
Need to have a test for this.

We should also probably update @deanpcmad's paddle gem to raise different types of errors so we can rescue from this specifically.

**Comment by nduplessis** [comment]
Yeah sorry about that, first time ever saving a PR as a draft and I didn't really expect it to show up here tbh. Will be adding a test for this soon and see if I can get a change to the paddle gem as suggested

**Comment by excid3** [comment]
No worries! You're good. Was mostly just making some notes. 👍

File: lib/pay/paddle_billing/billable.rb [hunk 1]
```diff
@@ -33,7 +33,13 @@ def customer
           sc
         end
       rescue ::Paddle::Error => e
-        raise Pay::PaddleBilling::Error, e
+        sc = Paddle::Customer.list(email:).data.first
```
Comment (excid3) [inline lib/pay/paddle_billing/billable.rb:36]: Paddle's CustomerAlreadyExists error message contains the ID in it, so we can parse it out of the string I think:

> "customer email conflicts with customer of id %s"

Comment (excid3) [reply lib/pay/paddle_billing/billable.rb:36]: Can always do that as a fallback if it can't find the value. 👍

**Comment by deanpcmad** [comment]
> We should also probably update @deanpcmad's paddle gem to raise different types of errors so we can rescue from this specifically.

Good idea 👍 
There are [quite a few](https://developer.paddle.com/errors/overview), so I won't want to manually add them all, but feel free to submit a PR for that

**Comment by shettytejas** [comment]
@nduplessis Is this PR still in development? This would be a great addition!

**Comment by nduplessis** [comment]
Hi yes it is, just had some deadlines I needed to get to. Will have this wrapped up before the end of the weekend!

**Comment by shettytejas** [comment]
@nduplessis That's great news! Thanks for the prompt reply!

**Comment by nduplessis** [comment]
@deanpcmad I've created a PR for this at https://github.com/deanpcmad/paddle/pull/6, let me know what you think of that approach please

**Comment by excid3** [comment]
This seems to have introduced a regression by changing the `def initialize(msg = nil)` signature for `Paddle::Error` as it now requires two arguments. Also Paddle Classic didn't get updated to match?

File: lib/pay/paddle_billing/billable.rb [hunk 2]
```diff
@@ -33,7 +33,13 @@ def customer
...
+        if sc.present?
+          pay_customer.update!(processor_id: sc.id)
```
Comment (excid3) [inline lib/pay/paddle_billing/billable.rb:38]: I'm not quite sure this logic makes sense. We would end up with Pay::Customer records with duplicate `processor_id` here, so we still need to raise an error here.

**Comment by deanpcmad** [comment]
With Paddle v2.5, it will now return a `Paddle::Errors::ConflictError` error when a customer already exists, so that can be integrated into this PR

**Comment by excid3** [comment]
Yeah, I'm just not sure what Pay should do in this case. Probably just continue raising the error as it currently does I guess?

**Comment by deanpcmad** [comment]
Yeah probably? Then a doc update to say it could return that error?

**Comment by excid3** [comment]
Yeah, probably want to just close this and mark it as a limitation in the Paddle overview docs. Same thing applies to Lemon Squeezy I'm pretty sure.

**Pull request closed without merging**